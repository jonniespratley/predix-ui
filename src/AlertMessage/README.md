The `<AlertMessage/>` component appear in a queue within the message container. Each message consists of an icon, title, body, and action.


## Usage

```react
<px.AlertMessage
  type='information'
  messageTitle='Alert'
  message='This is an information message'
  visible/>
```

> Note: This is an uncontrolled component.

### Types

```react
const {AlertMessage} = PxReact;
const types = ['info', 'warning', 'error', 'success', 'important', 'unknown'];
<div>
  <AlertMessage type='info' messageTitle='Alert' message='This is an message' visible/>
  <AlertMessage type='warning' messageTitle='Alert' message='This is an message' visible/>
  <AlertMessage type='error' messageTitle='Alert' message='This is an message' visible/>
  <AlertMessage type='important' messageTitle='Alert' message='This is an message' visible/>
  <AlertMessage type='success' messageTitle='Alert' message='This is an message' visible/>
  <AlertMessage type='unknown' messageTitle='Alert' message='This is an message' visible/>
</div>
```


## Properties

```table
span: 6
rows:
  - Name: action
    Type: String
    Description: Show dismiss icon, or show acknowledge button
  
  - Name: message
    Type: String
    Description: The content of the message.
  
  - Name: messageTitle
    Type: String
    Description: The title of the message.
  
  - Name: type
    Type: String
    Description: The type of message.
  
  - Name: actions
    Type: Node
    Description: Render actions component.
  
  - Name: onActionClick
    Type: Function
    Description: Invoked when action button is clicked
  
  - Name: onDismissClick
    Type: Function
    Description: Invoked when dismiss button is clicked

```



### Styling
The following custom properties are available for styling:
```table
span: 6
rows:
  - Property: --px-alert-message-background-color
    Default: null
    Description: Background color for the body of the alert

  - Property: --px-alert-message-text-color
    Default: null
    Description: Text color for the title and message

  - Property: --px-alert-message-color--important
    Default: null
    Description: Status color for an 'important' alert

  - Property: --px-alert-message-color--warning
    Default: null
    Description: Status color for a 'warning' alert

  - Property: --px-alert-message-color--error
    Default: null
    Description: Status color for an 'error' alert

  - Property: --px-alert-message-color--information
    Default: null
    Description: Status color for an 'information' alert

  - Property: --px-alert-message-dismiss-icon-color
    Default: null
    Description: Color of the dismiss icon
```
