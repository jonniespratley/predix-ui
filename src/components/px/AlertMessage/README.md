The `<AlertMessage/>` component appear in a queue within the message container. Each message consists of an icon, title, body, and action.

```hint
Reference https://www.predix-ui.com/#/elements/px-alert-message
```

## Usage

```react
<px.AlertMessage
  type='information'
  messageTitle='Alert'
  message='This is an information message'
  visible/>
```
```react
<px.AlertMessage
  type='error'
  messageTitle='Alert'
  message='This is an error message'
  visible/>
```

### Types

```react
<px.AlertMessage
  type='important'
  messageTitle='Alert'
  message='This is an important message'
  visible/>
```


## Properties

```table
span: 6
rows:
  - Name: action
    Type: string
    Description: dismiss, acknowledge
```



### Styling
The following custom properties are available for styling:
```table
span: 6
rows:
  - Property: --px-alert-message-background-color
    Default: null
    Description: Background color for the body of the alert

  - Property: --px-alert-message-text-color
    Default: null
    Description: Text color for the title and message

  - Property: --px-alert-message-color--important
    Default: null
    Description: Status color for an 'important' alert

  - Property: --px-alert-message-color--warning
    Default: null
    Description: Status color for a 'warning' alert

  - Property: --px-alert-message-color--error
    Default: null
    Description: Status color for an 'error' alert

  - Property: --px-alert-message-color--information
    Default: null
    Description: Status color for an 'information' alert

  - Property: --px-alert-message-dismiss-icon-color
    Default: null
    Description: Color of the dismiss icon
```
