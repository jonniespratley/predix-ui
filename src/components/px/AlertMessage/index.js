import React from 'react';
import styled, { css, keyframes } from 'styled-components';
import PropTypes from 'prop-types';
import Button from '../Button';

const fadeIn = keyframes`
  from {
    opacity: 0;
    left: 4rem;
  }
  to {
    opacity: 1;
    left: 0;
  }
`;
const fadeOut = keyframes`
  from {
    opacity: 1;
    max-height: 6.66667rem;
    min-height: 0px;
  }
  to {
    opacity: 0;
    max-height: 0px;
    min-height: 0px;
    padding-top: 0px;
    padding-bottom: 0px;
  }
`;

const AlertMessage = styled.div`
  visibility: hidden;
  display: none;
  width: var(--px-alert-message-width, auto);
  margin: 0.33333rem;
  padding: 0.66667rem;
  min-height: 4rem;
  word-wrap: break-word;
  word-break: break-word;
  background-color: var(--px-alert-message-background-color, white);
  position: relative;
  min-height: 0px;
  opacity: 0;
  max-height: 0px;
  padding-top: 0px;
  padding-bottom: 0px;
  box-shadow: 0px 4px 8px var(--px-shadow-notification, rgba(0, 0, 0, 0.2));
  animation: ${fadeOut} 0.4s cubic-bezier(0.78, 0.13, 0.16, 0.87);
  align-items: center;

  ${props => props.visible && css`
    overflow:hidden;
    height: 4rem;
    visibility: visible;
    display: flex;
    opacity: 1;
    max-height: none;
    padding-top: 0.66667rem;
    padding-bottom: 0.66667rem;
    animation: ${fadeIn} 0.4s cubic-bezier(0.78, 0.13, 0.16, 0.87);
  `}
  button {
    color: var(--px-alert-message-dismiss-icon-color, gray);
    outline: none;
    max-width: 50px;

    svg {
      height: 1.33333rem;
      width: 1.33333rem;
      fill: var(--px-alert-message-dismiss-icon-color, gray);
      stroke: var(--px-alert-message-dismiss-icon-color, gray);
    }
  }
`;

const AlertMessageContainer = styled.div`
  display: flex;
  align-items: center;
  flex: 2;
  margin-left: 0.66667rem;
  margin-right: 0.66667rem;
`;

const AlertMessageActions = styled.div`
  display: flex;
  align-items: center;
  position: relative;
  .dismiss {
    width: 44px;
    height: 44px;
    min-width: auto;
  }
`;

const AlertMessageSeverity = styled.div`
  min-height: 4rem;
  width: 4rem;
  margin-left: -0.6666rem;
  margin-top: -0.6666rem;
  margin-bottom: -0.6666rem;
  padding-top: 1rem;

  ${props => props.type === 'important' && css`
    background-color: var(--px-alert-message-color--important, red);
  `}

  ${props => (props.type === 'information' || props.type === 'info') && css`
    background-color: var(--px-alert-message-color--information, blue);
  `}

  ${props => props.type === 'warning' && css`
    background-color: var(--px-alert-message-color--warning, orange);
  `}

  ${props => props.type === 'error' && css`
    background-color: var(--px-alert-message-color--error, yellow);
  `}

  ${props => props.type === 'unknown' && css`
    background-color: var(--px-alert-message-color--unknown, gray);
  `}

  ${props => props.type === 'healthy' && css`
    background-color: var(--px-alert-label-background-color--healthy,green);
  `}

  ${props => props.type === 'success' && css`
    background-color: var(--px-alert-label-background-color--healthy,green);
  `}
`;

const Message = styled.div`
  color: var(--px-alert-message-text-color, black);
  line-height: 1.06667rem;
  font-size: 0.8rem;
  padding: 0;
  margin: 0;
`;

/**
 * AlertMessage component
 */
const Component = ({
  messageTitle,
  message,
  action,
  visible,
  actions,
  onActionClick,
  onDismiss,
  type,
  children
}) => (
  <AlertMessage visible={visible}>
    <AlertMessageSeverity type={type} />
    <AlertMessageContainer>
      <Message>
        <strong>{messageTitle} </strong>
        <span>{message}</span>
        <div>{children}</div>
      </Message>
    </AlertMessageContainer>
    <AlertMessageActions>
      {!actions && action === 'dismiss' &&
        <Button onClick={onDismiss} theme="bare" className="dismiss">
          <svg viewBox="0 0 22 22" preserveAspectRatio="xMidYMid meet" focusable="false">
            <g>
              <path strokeMiterlimit="10" d="M3 19L19 3M3 3l16 16" />
            </g>
          </svg>
        </Button>}

      {actions && actions()}

      {!actions && action === 'acknowledge' &&
        <Button
          id="actionButton"
          onClick={onActionClick}
          theme="tertiary"
        >OK
        </Button>}
    </AlertMessageActions>
  </AlertMessage>
);

Component.defaultProps = {
  action: null,
  type: 'information',
  message: null,
  messageTitle: null,
  visible: false,
  actions: null,
  children: null,
  onActionClick: null,
  onDismiss: null
};

Component.propTypes = {
  action: PropTypes.string,
  type: PropTypes.string,
  message: PropTypes.string,
  messageTitle: PropTypes.string,
  visible: PropTypes.bool,
  actions: PropTypes.func,
  onActionClick: PropTypes.func,
  onDismiss: PropTypes.func,
  children: PropTypes.node
};

Component.displayName = 'AlertMessage';

export default Component;
