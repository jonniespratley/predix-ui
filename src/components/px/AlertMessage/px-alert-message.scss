
   $inuit-enable-btn--bare: true;
   $inuit-enable-btn--tertiary: true;
   @import "px-buttons-design/_objects.buttons.scss";

   $actionable: true;
   @import "px-actionable-design/_objects.actionable.scss";
   @import "px-shadows-design/_objects.shadows.scss";
/**
 * px-alert-message styles
 */
 .px-alert-message {




   // Generic
   //@import "px-normalize-design/_generic.normalize.scss";
   //@import "px-box-sizing-design/_generic.box-sizing.scss";
   //@import "px-helpers-design/_generic.helpers.scss";

   // Base
   //@import "px-flexbox-design/_base.flexbox.scss";


   @keyframes fadein {
     from {
       opacity: 0;
       left: 4rem;
     }
     to {
       opacity: 1;
       left: 0;
     }
   }

   @keyframes fadeout {
     from {
       opacity: 1;
       max-height: $inuit-base-spacing-unit--small * 10;
       min-height: 0px;
     }
     to   {
       opacity: 0;
       max-height: 0px;
       min-height: 0px;
       padding-top: 0px;
       padding-bottom: 0px;
     }
   }

   // Component
   :root {
     --px-alert-label-background-color--important: white;
     --px-alert-label-text-color--important: black;
     --px-alert-label-background-color--error: white;
     --px-alert-label-text-color--error: black;
     --px-alert-label-background-color--warning: white;
     --px-alert-label-text-color--warning: black;
     --px-alert-label-background-color--information: white;
     --px-alert-label-text-color--information: black;
     --px-alert-label-background-color--unknown: white;
     --px-alert-label-text-color--unknown: black;
     --px-btn-min-width: 0
   }

   [hidden] {
     visibility: hidden;
     display: none;
   }

   .alert-message {
     visibility: hidden;
     display: none;
     width: $inuit-base-spacing-unit--small * 40;
     margin: $inuit-base-spacing-unit--tiny;
     padding: $inuit-base-spacing-unit--small;
     min-height: $inuit-base-spacing-unit--small * 6;
     word-wrap: break-word;
     word-break: break-word;
     background-color: var(--px-alert-message-background-color, white);
     position: relative;
     min-height: 0px;
     button, a {
         outline: none;
     }
     opacity: 0;
     max-height: 0px;
     padding-top: 0px;
     padding-bottom: 0px;
   }
   .alert-message--visible {
     visibility: visible;
     display: flex;
     opacity: 1;
     max-height: none;
     padding-top: $inuit-base-spacing-unit--small;
     padding-bottom: $inuit-base-spacing-unit--small;
   }
   .fade-in {
     animation: fadein .4s cubic-bezier(.78,.13,.16,.87);
   }
   .fade-out {
     animation: fadeout .4s cubic-bezier(.78,.13,.16,.87);
   }

   .message-column {
     display: flex;
     align-items: center;
     flex: 2;
     margin-left: $inuit-base-spacing-unit--small;
     margin-right: $inuit-base-spacing-unit--small;
   }

   .severity {
     min-height: 4rem;
     width: 4rem;
     margin-left: -0.6666rem; //10px
     margin-top: -0.6666rem;
     margin-bottom: -0.6666rem;
     padding-top: 1rem;
     &.important {
       background-color: var(--px-alert-message-color--important, red);
     }
     &.information {
       background-color: var(--px-alert-message-color--information, blue);
     }
     &.warning {
       background-color: var(--px-alert-message-color--warning, orange);
     }
     &.error {
       background-color: var(--px-alert-message-color--error, yellow);
     }
     &.unknown {
       background-color: var(--px-alert-message-color--unknown, gray);
     }
   }

   .message {
     color: var(--px-alert-message-text-color, black);
     line-height: calculateRem(16px);
     font-size: calculateRem(12px);
     padding: 0;
     margin: 0;

     &.collapsed-message{
       overflow: hidden;
       text-overflow: ellipsis;
       -webkit-box-orient: vertical;
       -webkit-line-clamp: 2;                /* number of lines to show */
       max-height: calculateRem(16px) * 2;   /* fallback */
     }
     &.more {
       font-size: 1.4rem;
       display: flex;
       justify-content: center;
       overflow: visible;
     }
     .title {
       font-weight: bold;
     }
   }
   .show-more {
     line-height: calculateRem(16px);
     font-size: calculateRem(12px);
   }

   .message-container {
     width: 100%;
   }

   .action {
     width: auto;

     cursor: pointer;
     margin-top: calculateRem(4px);
     .dismiss {
       color: var(--px-alert-message-dismiss-icon-color, gray);

       outline: none;
       max-width: 50px;

       svg {
         height: calculateRem(20px);
         width: calculateRem(20px);
         fill: var(--px-alert-message-dismiss-icon-color, gray);
         stroke: var(--px-alert-message-dismiss-icon-color, gray);
       }
     }
   }

 }
