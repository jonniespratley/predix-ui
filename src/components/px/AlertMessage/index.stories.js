import React from 'react';
import { storiesOf } from '@storybook/react';

//addons
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import { withKnobs, text, array, object, boolean, number, select } from '@storybook/addon-knobs';

import AlertMessage from './';

const typeOptions = {
  information: 'Info',
  error: 'Error',
  important: 'Important',
  healthy: 'Healthy',
  warning: 'Warning',
  success: 'Success'
};
const navItems = [
  {
    "path": "/page1",
    "icon": "px-fea:analysis",
    "label": "Page 1"
  },
  {
    "path": "/page2",
    "icon": "px-fea:dashboard",
    "label": "Page 2"
  }
];
///
storiesOf('Alert Message', module)
  .addDecorator(withKnobs)
	.add('default', () => (
		<AlertMessage
      onDismissClick={action('onDismissClick')}
      onActionClick={action('onActionClick')}
      type={select('type', Object.keys(typeOptions))}
      action={select('action', ['acknowledge', 'dismiss'])}
      message={text('message', 'This is the message content.')}
      messageTitle={text('messageTitle', 'Alert')}
      visible={boolean('visible', true)}/>
  ))
  .add('with message', () => (
		<AlertMessage
      messageTitle='Alert'
      message='This is an information message'
      visible/>
  ))
  .add('with type "information"', () => (
		<AlertMessage
      messageTitle='Alert'
      message='This is a message'
      type='information'
      visible/>
  ))
  .add('with type "important"', () => (
		<AlertMessage
      messageTitle='Alert'
      message='This is a message'
      type='important'
      visible/>
  ))
  .add('with type "error"', () => (
		<AlertMessage
      messageTitle='Alert'
      message='This is a message'
      type='error'
      visible/>
  ))
  .add('with type "warning"', () => (
		<AlertMessage
      messageTitle='Alert'
      message='This is a message'
      type='warning'
      visible/>
  ))
  .add('with type "unkown"', () => (
		<AlertMessage
      messageTitle='Alert'
      message='This is a message'
      type='unkown'
      visible/>
	))
	;
