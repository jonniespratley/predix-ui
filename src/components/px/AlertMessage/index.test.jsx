//import { expect } from 'chai';
import React from 'react';
import { shallow } from 'enzyme';
import AlertMessage from './';
import renderer from 'react-test-renderer';

const typeOptions = {
  information: 'Info',
  error: 'Error',
  important: 'Important',
  healthy: 'Healthy',
  warning: 'Warning',
  success: 'Success'
};

describe('AlertMessage', () => {

  test('matches snapshot', () => {
    const tree = renderer.create(<AlertMessage type='important' visible />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('display', 'flex');
    //expect(tree).toHaveStyleRule('color', 'var(--px-alert-label-text-color--information,white)');
  });
  test('hidden - matches snapshot', () => {
    const tree = renderer.create(<AlertMessage type='important' />).toJSON();
    expect(tree).toMatchSnapshot();

  });

  Object.keys(typeOptions).forEach(val => {
    test(`${val} matches snapshot`, () => {
      const tree = renderer.create(<AlertMessage type={val} visible>This is a message</AlertMessage>).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
