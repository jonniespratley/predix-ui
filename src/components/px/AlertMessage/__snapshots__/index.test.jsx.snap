// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertMessage error matches snapshot 1`] = `
<div
  className="ebxlRt"
>
  <div
    className="fFxuom"
    type="error"
  />
  <div
    className="bwRksd"
  >
    <div
      className="etuWhZ"
    >
      <strong>
         
      </strong>
      <span />
      <div>
        This is a message
      </div>
    </div>
  </div>
  <div
    className="ipGZxu"
  />
</div>
`;

exports[`AlertMessage healthy matches snapshot 1`] = `
<div
  className="ebxlRt"
>
  <div
    className="iBcrDo"
    type="healthy"
  />
  <div
    className="bwRksd"
  >
    <div
      className="etuWhZ"
    >
      <strong>
         
      </strong>
      <span />
      <div>
        This is a message
      </div>
    </div>
  </div>
  <div
    className="ipGZxu"
  />
</div>
`;

exports[`AlertMessage hidden - matches snapshot 1`] = `
<div
  className="hkbWnA"
>
  <div
    className="lgYmRs"
    type="important"
  />
  <div
    className="bwRksd"
  >
    <div
      className="etuWhZ"
    >
      <strong>
         
      </strong>
      <span />
      <div />
    </div>
  </div>
  <div
    className="ipGZxu"
  />
</div>
`;

exports[`AlertMessage important matches snapshot 1`] = `
<div
  className="ebxlRt"
>
  <div
    className="lgYmRs"
    type="important"
  />
  <div
    className="bwRksd"
  >
    <div
      className="etuWhZ"
    >
      <strong>
         
      </strong>
      <span />
      <div>
        This is a message
      </div>
    </div>
  </div>
  <div
    className="ipGZxu"
  />
</div>
`;

exports[`AlertMessage information matches snapshot 1`] = `
<div
  className="ebxlRt"
>
  <div
    className="lhPWtr"
    type="information"
  />
  <div
    className="bwRksd"
  >
    <div
      className="etuWhZ"
    >
      <strong>
         
      </strong>
      <span />
      <div>
        This is a message
      </div>
    </div>
  </div>
  <div
    className="ipGZxu"
  />
</div>
`;

exports[`AlertMessage matches snapshot 1`] = `
<div
  className="ebxlRt"
>
  <div
    className="lgYmRs"
    type="important"
  />
  <div
    className="bwRksd"
  >
    <div
      className="etuWhZ"
    >
      <strong>
         
      </strong>
      <span />
      <div />
    </div>
  </div>
  <div
    className="ipGZxu"
  />
</div>
`;

exports[`AlertMessage success matches snapshot 1`] = `
<div
  className="ebxlRt"
>
  <div
    className="EfQEb"
    type="success"
  />
  <div
    className="bwRksd"
  >
    <div
      className="etuWhZ"
    >
      <strong>
         
      </strong>
      <span />
      <div>
        This is a message
      </div>
    </div>
  </div>
  <div
    className="ipGZxu"
  />
</div>
`;

exports[`AlertMessage warning matches snapshot 1`] = `
<div
  className="ebxlRt"
>
  <div
    className="kZaSPE"
    type="warning"
  />
  <div
    className="bwRksd"
  >
    <div
      className="etuWhZ"
    >
      <strong>
         
      </strong>
      <span />
      <div>
        This is a message
      </div>
    </div>
  </div>
  <div
    className="ipGZxu"
  />
</div>
`;
